discovery/balancer_builder.go

```diff
@@ -0,0 +1,60 @@
+package discovery
+
+import (
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-hclog"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+)
+
+// balancerBuilder is invoked by gRPC to construct our custom balancer. This
+// handles the hook up of the balancer to a Watcher.
+type balancerBuilder struct {
+	watcher *Watcher
+	log     hclog.Logger
+
+	// This is the unique id for this builder in gRPC global balancer registry,
+	// used as the loadBalancingPolicy name when configuring a gRPC connection.
+	hackPolicyName string
+}
+
+var _ balancer.Builder = (*balancerBuilder)(nil)
+
+// registerBalancer creates new balancer.Builder for the given Watcher and registers it with gRPC.
+// Balancers must be globally registered and cannot be unregistered, so we register each builder
+// with a unique id to facilitate testing. This returns the loadBalancingPolicy name used to enable
+// this balancer on a gRPC connection.
+func registerBalancer(w *Watcher, log hclog.Logger) string {
+	b := &balancerBuilder{
+		watcher:        w,
+		log:            log,
+		hackPolicyName: uuid.New().String(),
```

Comment:
This is a total nit so please feel encouraged to ignore it entirely 😅 

As we don't need a UUID per se, just some random string, we could hex-encode random bytes from the `math/rand` stdlib package to avoid adding a dependency.

```Go
func randomString() string {
    b := make([]byte, 32)
    if _, err := rand.Read(b); err != nil {
        panic("failed to generate random bytes")
    }
    return hex.EncodeToString(b)
}
```